R/modify-recursive.R

```diff
@@ -0,0 +1,68 @@
+#' Recursively modify a list
+#'
+#' `rmodify()` allows you to recursively modify a list, supplying functions
+#' that either modify each leaf or each node (or both).
+#'
+#' @param x A list.
+#' @param f_leaf A function applied to each leaf.
+#' @param p_leaf A predicate function that returns `TRUE` when an element is
+#'   a leaf, determining whether `f_leaf` or `f_pre`/`f_post` is applied to
+#'   it. The default value, `NULL`, treats lists as nodes and everything else
+#'   as leaves.
+#' @param f_pre,f_post Functions applied to each node. `f_pre` is applied
+#'   the tree is traversed "down", i.e. before the leaves are transformed
+#'   with `f_leaf`, while `f_post` is applied on the way "up", i.e.
+#'   after the leaves are transformed.
+#' @family map variants
+#' @export
+#' @examples
+#' x <- list(list(a = 2:1, c = list(b1 = 2), b = list(c2 = 3, c1 = 4)))
+#' x |> str()
+#'
+#' # Transform each leaf
+#' x |> rmodify(f_leaf = \(x) x + 100) |>  str()
+#'
+#' # Recursively sort the nodes
+#' sort_named <- function(x) {
+#'   nms <- names(x)
+#'   if (!is.null(nms)) {
+#'     x[order(nms)]
+#'   } else {
+#'     x
+#'    }
+#' }
+#' x |> rmodify(f_post = sort_named) |> str()
+rmodify <- function(x,
+                    f_leaf = identity,
+                    p_leaf = NULL,
+                    f_pre = identity,
+                    f_post = identity) {
+  if (!is_vector(x)) {
+    cli::cli_abort("{.arg x} must be a vector, not {.obj_type_friendly {x}}.")
+  }
+
+  f_post <- rlang::as_function(f_post)
+  f_pre <- rlang::as_function(f_pre)
+  f_leaf <- rlang::as_function(f_leaf)
+  if (is.null(p_leaf)) {
+    is_leaf <- function(x) {
+      !is.list(x)
```

Comment:
I would have expected something like `vec_is_list()` here.

Reply:
Implying we might need some tests for determining whether a data frame is a node or a leaf

Reply:
My gut feeling is that `is.list` is right here, but I can't articulate why and if you both think `vec_is_list()` is correct then that's probably right.

Reply:
Perhaps because of Vitalie's example about stripping environments of formulas? That's a case where you want `modify_tree()` to reach everywhere. Should we export a predicate for this case, i.e. to reach everywhere `[[<-` can? But I think the default should be to only descend explicit lists.

Reply:
We're generally switching to `vec_is()` and `vec_is_list()` to determine scalarness and listness.

Reply:
@lionel- generally, yes, but not that much inside of purrr to maintain backward compatibility. I think the main question here is it better for purrr to be internally consistent or for newer functions to be consistent with newer vctrs standards. I suspect package consistency is more important.

(That said, I think @vspinu argued that `list_modify()` shouldn't recurse into data frames)

Reply:
In the interests of simplicity, lets stick with `is.list` here. I do expect that many people will need to customise `is_leaf` so it's ok for the default to have limited use.